### ISSUE TITLE
odd gridline behavior when redrawing chart (non-idempotent?)

### ISSUE THREAD

**Issue by eponymous301** [d3fc/d3fc#1087]
I'm trying to create a candlestick chart that can be scrolled by keyboard, and I redraw the chart by repeating the original drawing code with different array of data.

On some dates when there are large jumps in range I get strange non-even spacing of grid lines:

![screen shot 2017-10-12 at 4 32 58 pm](https://user-images.githubusercontent.com/29875759/31523879-0b43788a-af6b-11e7-8c6d-ff88b470c676.png)

![screen shot 2017-10-12 at 4 28 53 pm](https://user-images.githubusercontent.com/29875759/31523787-82a0593a-af6a-11e7-9abd-fa1fbf97fe75.png)

I've posted a replication case on codepen: [https://codepen.io/eponymous/pen/mBjozV](https://codepen.io/eponymous/pen/mBjozV) - click on the lower pane where the graph is, then hit the 'left' key twice. Hitting 'right' and 'left' to scroll the lowest (green) candlestick on and off the screen will reliably replicate the issue.

 

It seems to happen only when drawing using a div that is not empty due to previous call. The problem goes away if I delete all the children of the div before redrawing the chart, but this slows things down and causes flicker.

**Comment by chrisprice** [comment]
Seems related to the scale caching in the gridlines component, hacking the [line](https://github.com/d3fc/d3fc-annotation/blob/master/src/svg/gridline.js#L64)
```
const yScaleOld = element.__y_scale__ || yScale;
```
to be
```
const yScaleOld = yScale;
```
seems to stop it happening. 

Unfortunately that line is required for transition support. Needs further investigation.

**Comment by ColinEberhardt** [comment]
For reference, that behaviour was derived from d3-axis:

https://github.com/d3/d3/blob/v3.5.17/src/svg/axis.js#L25

And is also present in d3fc-axis:

https://github.com/d3fc/d3fc-axis/blob/master/src/axis.js#L74

**Comment by eponymous301** [comment]
Thank you Chris, I am not using transitions so the hack may be all I need.

To eliminate flicker I attempted another workaround with frame buffering using two divs, drawing to a hidden div and swapping, but I ran into another complication, that d3 apparently draws asynchronously, and my subsequent div hide/show code ran before the chart was drawn:

`    d3.select(hidden_div_id).datum(data).call(chart); ` 
`      switch_visible_div(); ` 


 I searched for a way to add an 'afterDraw' callback but this seems to be an open request in d3. Is there some way I've missed to either add a callback or get d3 to draw synchronously?
